src/enclosure.c

```diff
@@ -171,3 +184,142 @@ int enclosure_open(const struct enclosure_device *enclosure)
 
 	return fd;
 }
+
+static struct enclosure_device *find_enclosure(char *enclosure_id)
+{
+	struct enclosure_device *enclosure_device;
+	list_for_each(sysfs_get_enclosure_devices(), enclosure_device) {
+		if (strncmp(enclosure_device->dev_path, enclosure_id, PATH_MAX) == 0) {
+			return enclosure_device;
+		}
+	}
+	return NULL;
+}
+
+static struct ses_slot *find_enclosure_slot_by_index(struct enclosure_device *encl, int index)
+{
+	for (int i = 0; i < encl->slots_count; i++) {
+		if (encl->slots[i].index == index) {
+			return &encl->slots[i];
+		}
+	}
+	return NULL;
+}
+
+static status_t _enclosure_get_slot(struct enclosure_device *encl, int index, struct slot_response *slot_res)
+{
+	struct block_device *block_device = NULL;
+	struct ses_slot *s_slot = find_enclosure_slot_by_index(encl, index);
+	if (!s_slot) {
+		log_error("SCSI: Unable to locate slot in enclosure %d\n", index);
+		return STATUS_NULL_POINTER;
+	}
+
+	slot_res->state = s_slot->ibpi_status;
+	snprintf(slot_res->slot, PATH_MAX, "%s/%d", encl->dev_path, index);
+
+	// Not having a block device for a slot is OK
+	if (NULL != (block_device = locate_block_by_sas_addr(s_slot->sas_addr))) {
+		snprintf(slot_res->device, PATH_MAX, "/dev/%s", basename(block_device->sysfs_path));
+	}
+	return STATUS_SUCCESS;
+}
+
+static status_t parse_slot_num(char *slot_num, char *enclosure_id, int *index)
+{
+	char tmp_enclosure_id[PATH_MAX];
+	const char *index_str;
+
+	if (slot_num == NULL || slot_num[0] == '\0') {
+		log_error("SCSI: Invalid slot identifier =%s\n", slot_num);
+		return STATUS_NULL_POINTER;
+	}
+
+	snprintf(tmp_enclosure_id, PATH_MAX, "%s", slot_num);
+	snprintf(enclosure_id, PATH_MAX, "%s", dirname(tmp_enclosure_id));
+	index_str = basename(slot_num);
+
+	if (str_toi(index, index_str, NULL, 10)) {
+		log_error("SCSI: Invalid slot identifier index %s\n", index_str);
+		return STATUS_INVALID_PATH;
+	}
+	return STATUS_SUCCESS;
+}
+
+static status_t enclosure_get_slot_by_slot_num(char *slot_num, struct slot_response *slot_res)
+{
+	char enclosure_id[PATH_MAX];
+	int index = -1;
+	struct enclosure_device *encl = NULL;
+
+	status_t parse = parse_slot_num(slot_num, enclosure_id, &index);
+	if (STATUS_SUCCESS != parse)
+		return parse;
+
+	encl = find_enclosure(enclosure_id);
+	if (!encl) {
+		log_error("SCSI: Invalid enclosure ='%s'\n", enclosure_id);
+		return STATUS_NULL_POINTER;
+	}
+	return _enclosure_get_slot(encl, index, slot_res);
+}
+
+static status_t enclosure_get_slot_by_device(char *device, struct slot_response *slot_res)
+{
+	struct block_device *block_device = get_block_device_from_sysfs_path(basename(device));
+
+	if (!block_device) {
+		log_error("SCSI: Device node not found %s\n", device);
+		return STATUS_INVALID_PATH;
+	}
+
+	if (!block_device->enclosure) {
+		log_error("SCSI: Not a SCSI ses device %n\n", device);
+		return STATUS_INVALID_PATH;
+	}
+	return _enclosure_get_slot(block_device->enclosure, block_device->encl_index, slot_res);
```

Comment:
> maybe let's set `slot_res->device` here

Please clarify what you mean by this.  We will still need to assign `slot_res->device` in the `_enclosure_get_slot` as it's needed in the `enclosure_get_slot_by_slot_num` path?

Reply:
Please take a look at https://github.com/intel/ledmon/pull/109/commits/3baf05cf43e372ffb49d14858a946bf9ef2388b3 for what I think you might be thinking.

Reply:
Yeah, I was thinking about something like that
However, please change it, so that device is passed as optional parameter to `_enclosure_get_slot()`
So that setting `slot_res` values is done only in `_enclosure_get_slot()`